File: 3Sum_Closest.java
/*16. 3Sum Closest Add to List
Description  Submission  Solutions
Total Accepted: 116293
Total Submissions: 379047
Difficulty: Medium
Contributors: Admin
Given an array S of n integers, find three integers in S such that the sum is closest to a given number, target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

    For example, given array S = {-1 2 1 -4}, and target = 1.

    The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
Hide Company Tags Bloomberg
Hide Tags Array Two Pointers
Hide Similar Problems (M) 3Sum (M) 3Sum Smaller
 */






/*
1. sort array
2. 3 sum to solve question
    each time check if current sum is closer than closeSum
 */

////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public int threeSumClosest(int[] nums, int target) {
        if(nums==null || nums.length==0) return 0;
        Arrays.sort(nums);
        int closeSum = nums[0] + nums[1] + nums[2];
        for(int i = 0; i < nums.length-2; ++i){
            int low = i+1;
            int high = nums.length-1;
            while(low < high){
                int sum = nums[i] + nums[low] + nums[high];
                if(sum > target){
                    high--;
                } else {
                    low++;
                }
                if(Math.abs(sum-target) < Math.abs(closeSum-target)) closeSum = sum;
            }
        }
        return closeSum;
    }
}

File: 3Sum_Closest.swift
/*
Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target. Return the sum of the three integers. You may assume that each input would have exactly one solution.



Example 1:

Input: nums = [-1,2,1,-4], target = 1
Output: 2
Explanation: The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).


Constraints:

3 <= nums.length <= 10^3
-10^3 <= nums[i] <= 10^3
-10^4 <= target <= 10^4
*/

/*
Solution 1:
sort + 2 pointer

sort array
fix first element, then use two pointer to get close to target

Time Complexity: O(n^2)
Space Complexity: O(n)
*/
class Solution {
    func threeSumClosest(_ nums: [Int], _ target: Int) -> Int {
        let n = nums.count
        var nums = nums.sorted()
        var sum = nums[0] + nums[1] + nums[2]

        for i in 0..<(n-2) {
            var left = i+1
            var right = n-1
            while left < right {
                var temp = nums[i] + nums[left] + nums[right]
                if temp > target {
                    right -= 1
                } else {
                    left += 1
                }
                if abs(temp - target) < abs(sum - target) {
                    sum = temp
                }
            }
        }
        return sum
    }
}